pkg/library/dag/dag.go

```diff
@@ -0,0 +1,283 @@
+package dag
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/kuadrant/kuadrant-operator/pkg/library/utils"
+)
+
+// A Directed Acyclic Graph (DAG) is a graph representing a structure formed by vertices, or nodes,
+// connected by directed edges.
+// In a DAG, each edge has an initial node, called the parent, and a final node, called the child.
+// The graph is considered acyclic because it does not contain any cycles,
+// meaning there are no sequences of consecutive directed edges that form a closed loop.
+
+type NodeID = string
+
+type nodeNotFoundError struct {
+	id NodeID
+}
+
+func (e *nodeNotFoundError) Error() string {
+	return fmt.Sprintf("node %s not found", e.id)
+}
+
+func IsNodeNotFound(err error) bool {
+	var nodeNotFoundErr *nodeNotFoundError
+	return errors.As(err, &nodeNotFoundErr)
+}
+
+type Node interface {
+	ID() NodeID
+}
+
+type internalNode struct {
+	id       NodeID
+	node     Node
+	parents  map[NodeID]*internalNode
+	children map[NodeID]*internalNode
```

Comment:
Should we use [`orderedmap`](https://pkg.go.dev/github.com/elliotchance/orderedmap/v2) here, so we have consistent responses from funcs like `Parents()`, `Children()`?

Reply:
happy to add them, but it is not needed right now, is it?

Reply:
This is a basic data structure that can be used to build indexes and elaborate logic on top. Because we already had problems with that in the past, I'm personally more inclined every day to ensuring consistent ordering on those lists. But up to you.